### ISSUE TITLE
Convert Signature and Public key into G1 and G2 format.

### ISSUE THREAD

**Issue by wdcs-amitkumar** [maidsafe/bls_dkg#119]
Hi Team !!
Thanks you bls-dkg to provide such a complex implemetation in rust.

I would like to request if it is possible to convert public key and signature into G1 and G2 format.
So that I can verify dkg-bls signature in the ethereum smart contract.

Currently I tried to input those PK and SIG in [python ECC library ](https://github.com/ethereum/py_ecc/blob/master/py_ecc/bls/g2_primitives.py)
**Line 40 , 51** 

but it produces G1 and G2 of length 115 but [this smart contract ](https://gist.github.com/BjornvdLaan/ca6dd4e3993e1ef392f363ec27fe74c4#file-blsexample-sol) Line 38 **verifyBLSTest()**

accepts G1 G2 value of length 77.

In short do any way to get G1 G2 values in for required length ??

**Comment by wdcs-amitkumar** [comment]
Meanwhile I tried to convert public key and signature into G1 and G2 and found that
They are producing output of different length.

![Screenshot from 2022-01-10 02-27-34](https://user-images.githubusercontent.com/62380569/148742896-eb9c3930-2e6d-4372-986e-85966965208f.png)

![image](https://user-images.githubusercontent.com/62380569/148742975-c9739801-6fe0-4380-acab-abd9a0a2379c.png)

**Comment by wdcs-amitkumar** [comment]
[BLS_VERIFY_ETH.pdf](https://github.com/maidsafe/bls_dkg/files/7837823/BLS_VERIFY_ETH.pdf)

Just for reference of Signature I generated from BLS DKG rust and Python

**Comment by amitcz** [comment]
I have found this one but not sure it is similar 
```
fn main() {

    let priv_key = bls_signatures::PrivateKey::new(&[0_u8; 32]);  // not secure, but repeatable

    let affine = priv_key
        .public_key()
        .as_affine(); // blstrs::G1Affine, no longer from paired!


    println!("{:?}", affine.x());
    println!("{:?}", affine.y());
}
```

**Comment by maqi** [comment]
Hi, @amitcz,

Thank you for the comment.

First, regarding the table of PK/Sig length you attached above, I think for bls_dkg, it shall be 192 for the PK and 384 for the Sig ?

Second, the format of PK/Sig that bls_dkg using is from [threshold_crypto](https://docs.rs/threshold_crypto/0.4.0/threshold_crypto/) .
In that crate, you can define the length of PK/Sig by changing the value of PK_SIZE and SIG_SIZE.
That crate has test using mock PK/Sig type by change the PK_SIZE and SIG_SIZE to 4.
I am not sure whether the bls_dkg can support a changed PK_SZIE/SIG_SIZE directly, worth to have a try.

Hope this helps.
Cheers,

**Comment by davidrusu** [comment]
Just to add some notes here,

@amitcz there are a few different representations of G1/G2 points and there's a distinction between PublicKeySets and PublicKeys that you may missing.

G1/G1 can be in compressed or affine forms (or projective, but this is rarely serialized), these sizes are in the table below
```
             | G1 |  G2 |
-------------------------
| compressed | 48 |  96 |
| affine     | 96 | 192 |
```

Make sure you are converting your G1/G2 points into the representation your function expects.

Second, you may be confusing the PublicKeySet for the PublicKey. make sure to call [public_key()](https://maidsafe.github.io/bls_dkg/bls_dkg/key_gen/struct.PublicKeySet.html#method.public_key)

BLS-DKG exposes the compressed G1/G2 points, by calling `to_bytes()` on the `PublicKey` / `PublicKeyShare` / `Signature` / `SignatureShare`

This compressed representation should be parse-able by other BLS libraries and from that you can convert to Affine if necessary.

**Comment by amitcz** [comment]
> I have found this one but not sure it is similar
> 
> ```
> fn main() {
> 
>     let priv_key = bls_signatures::PrivateKey::new(&[0_u8; 32]);  // not secure, but repeatable
> 
>     let affine = priv_key
>         .public_key()
>         .as_affine(); // blstrs::G1Affine, no longer from paired!
> 
> 
>     println!("{:?}", affine.x());
>     println!("{:?}", affine.y());
> }
> ```

https://github.com/maidsafe/bls_dkg/blob/58e26b0d33b501912effc4e571f93a80d56624a5/src/key_gen/tests.rs#L273

This actually prints 384 char Public key and 192 char Signature..

**Comment by wdcs-amitkumar** [comment]
> Hi, @amitcz,
> 
> Thank you for the comment.
> 
> First, regarding the table of PK/Sig length you attached above, I think for bls_dkg, it shall be 192 for the PK and 384 for the Sig ?
> 
> Second, the format of PK/Sig that bls_dkg using is from [threshold_crypto](https://docs.rs/threshold_crypto/0.4.0/threshold_crypto/) . In that crate, you can define the length of PK/Sig by changing the value of PK_SIZE and SIG_SIZE. That crate has test using mock PK/Sig type by change the PK_SIZE and SIG_SIZE to 4. I am not sure whether the bls_dkg can support a changed PK_SZIE/SIG_SIZE directly, worth to have a try.
> 
> Hope this helps. Cheers,

https://github.com/maidsafe/bls_dkg/blob/58e26b0d33b501912effc4e571f93a80d56624a5/src/key_gen/tests.rs#L273
This line prints public key : 384 characters and Signature : 192 characters
What encoding has been used ??

**Comment by wdcs-amitkumar** [comment]
Hi Team Maidsafe!!

Thank you for your efforts, I am able to get Affine values for my public key and signature.
Here is the code with changes : [](https://pastebin.com/EaYppJuL)

![Screenshot from 2022-01-11 02-15-08](https://user-images.githubusercontent.com/62380569/148836665-70940686-1d67-4614-8d73-c1e3f74aadbf.png)

**One last help I need** 
How can i convert those hex affine values into big int that following contract can take as input 

Line 42 :[ bls-verifier-contract]( https://gist.github.com/BjornvdLaan/ca6dd4e3993e1ef392f363ec27fe74c4#file-blsexample-sol)

Again very much thankful for all your support.
Waiting for response :)

**Comment by davidrusu** [comment]
I don't think we can help you much with your contract, but those numbers are hex encoded, if you want the decimal representation, you can simply re-encode as decimal

**Comment by wdcs-amitkumar** [comment]
Hi @davidrusu 

I tried to convert public key points , it is giving 115 character numbers, 
But smart contract (since it supports max uin256 size) only takes 77 character number as input.

Still hopeful 

It it possible to get 77 character long public key from your repo.

Appreciated if provided with code snippet.
Thanks :)

**Comment by davidrusu** [comment]
PublicKeys are always 48 bytes when compressed and 96 bytes in Affine.

>  it is giving 115 character numbers,

I'm not sure what a "character" is, but you should be counting in bytes, not characters.

**Comment by maqi** [comment]
> It it possible to get 77 character long public key from your repo.

You may try with the threshold_crypto repo to refactor the PK_SIZE and SIG_SIZE to what you want.
However, this may comes with cost of reduced security.
